msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -1064,51 +1064,174 @@ protected void acquireTokenSilent(@NonNull final String[] scopes,
+            // use home account - b2c is not compatible with broker, so no need to construct
+            // the account used in the request...
+            return AccountAdapter.getAccountInternal(
+                    mPublicClientConfiguration.getClientId(),
+                    mPublicClientConfiguration.getOAuth2TokenCache(),
+                    multiTenantAccount.getHomeAccountId(),
+                    multiTenantAccount.getTenantId()
+            );
+        } else if (authority instanceof AzureActiveDirectoryAuthority) {
+            final AzureActiveDirectoryAuthority aadAuthority = (AzureActiveDirectoryAuthority) authority;
 
-        try {
-            final TokenCommand silentTokenCommand = new TokenCommand(
-                    params,
-                    MSALControllerFactory.getAcquireTokenSilentControllers(
-                            mPublicClientConfiguration.getAppContext(),
-                            params.getAuthority(),
-                            mPublicClientConfiguration
-                    ),
-                    callback
+            // Although the below call implies the returned value will be a tenantId, this isn't
+            // strictly true: it can also be an alias such as 'common', 'consumers', or
+            // 'organizations'. Additionally, if the developer has used a named tenant it could end
+            // up being something like <tenant_name>.onmicrosoft.com
+            //
+            // If the tenant is a GUID, we need to choose the account/profile that corresponds to it
+            //
+            // If the tenant is named 'common', 'consumers', or 'organizations' we need to use the
+            // home account
+            //
+            // If the account is named like <tenant_name>.onmicrosoft.com we need to query the OpenId
+            // Provider Configuration Metadata in order to get the tenant id. Once we have the
+            // tenant id, we must then select the appropriate home or profile.
+            String tenantIdNameOrAlias = aadAuthority.getAudience().getTenantId();
+
+            // The AccountRecord we'll use to request a token...
+            final AccountRecord accountRecord = new AccountRecord();
+            accountRecord.setEnvironment(multiTenantAccount.getEnvironment());
+            accountRecord.setHomeAccountId(multiTenantAccount.getHomeAccountId());
+
+            final boolean isUuid = isUuid(tenantIdNameOrAlias);
+
+            if (!isUuid && !isHomeTenantEquivalent(tenantIdNameOrAlias)) {
+                final OpenIdProviderConfiguration providerConfiguration =
+                        loadOpenIdProviderConfigurationMetadata(requestAuthority);
+
+                final String issuer = providerConfiguration.getIssuer();
+                final Uri issuerUri = Uri.parse(issuer);
+                final List<String> paths = issuerUri.getPathSegments();
+                tenantIdNameOrAlias = paths.get(0);
```

Comment:
Should we check for size of paths list here?

Reply:
This result will follow a standard format of `https://login.microsoftonline.com/{tenantid}/v2.0` - if it doesn't this would be problematic. I can add a check here, though there isn't anything we can do other than throw an Exception

Reply:
Added a simple test in a472a57e92cc5edf72b52aa672d70faddff76c9c - not exhaustive (for instance, doesn't assert that the value is actually a `UUID`) but I think it suffices